[Bug]: AttributeError: 'NWBHDF5IO' object has no attribute '_HDF5IO__open_links'

### What happened?

@TheChymera managed to get our CIs all (seems happens across OSes) fail out with

```
FAILED dandi/tests/test_files.py::test_validate_bogus - pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7f42d37f4550>

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/hdmf/backends/io.py", line 137, in __del__
    self.close()
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 738, in close
    self.close_linked_files()
  File "/opt/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 749, in close_linked_files
    for obj in self.__open_links:
AttributeError: 'NWBHDF5IO' object has no attribute '_HDF5IO__open_links'
```
which you can observe in https://github.com/dandi/dandi-cli/actions/runs/3952213342/jobs/6767014502   . What could this be due to???

### Steps to Reproduce

```shell
- Get born to be @TheChymera
- Send a PR
```


### Traceback

```shell
above
```


### Operating System

Linux

### Python Executable

Conda

### Python Version

3.10

### Package Versions

2023-01-18T19:23:49.7460559Z Successfully installed Deprecated-1.2.13 SecretStorage-3.3.3 anys-0.2.1 appdirs-1.4.4 arrow-1.2.3 asciitree-0.3.3 attrs-22.2.0 bidsschematools-0.6.0 blessings-1.7 certifi-2022.12.7 cffi-1.15.1 charset-normalizer-3.0.1 ci-info-0.3.0 click-8.1.3 click-didyoumean-0.3.0 coverage-7.0.5 cryptography-39.0.0 dandi-0.48.1+12.g65368a6 dandischema-0.7.1 dnspython-2.3.0 email-validator-1.3.0 entrypoints-0.4 etelemetry-0.3.0 exceptiongroup-1.1.0 fasteners-0.18 fqdn-1.5.1 fscacher-0.2.0 h5py-3.7.0 hdmf-3.5.0 humanize-4.4.0 idna-3.4 importlib-metadata-6.0.0 iniconfig-2.0.0 interleave-0.2.1 isodate-0.6.1 isoduration-20.11.0 jaraco.classes-3.2.3 jeepney-0.8.0 joblib-1.2.0 jsonpointer-2.3 jsonschema-4.17.3 keyring-23.13.1 keyrings.alt-4.2.0 more-itertools-9.0.0 natsort-8.2.0 numcodecs-0.11.0 numpy-1.23.5 nwbinspector-0.4.25 opencv-python-4.7.0.68 packaging-23.0 pandas-1.5.2 pluggy-1.0.0 pycparser-2.21 pycryptodomex-3.16.0 pydantic-1.10.4 pynwb-2.2.0 pyout-0.7.2 pyrsistent-0.19.3 pytest-7.2.1 pytest-cov-4.0.0 pytest-mock-3.10.0 python-dateutil-2.8.2 pytz-2022.7.1 pyyaml-6.0 requests-2.28.2 responses-0.22.0 rfc3339-validator-0.1.4 rfc3987-1.3.8 ruamel.yaml-0.17.21 ruamel.yaml.clib-0.2.7 scipy-1.10.0 semantic-version-2.10.0 six-1.16.0 tenacity-8.1.0 toml-0.10.2 tomli-2.0.1 tqdm-4.64.1 types-toml-0.10.8.1 typing-extensions-4.4.0 uri-template-1.2.0 urllib3-1.26.14 webcolors-1.12 wrapt-1.14.1 zarr-2.13.6 zipp-3.11.0

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/hdmf-dev/hdmf/blob/dev/.github/CODE_OF_CONDUCT.md)
- [X] Have you checked the [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this bug was not already [reported](https://github.com/hdmf-dev/hdmf/issues)?

So I also tested this with different versions, and apparently the issue shows up with `=hdmf-3.5.0` but not with `=hdmf-3.4.7`.

Sadly, it also still happens after applying the proposed patch :(

```
(dev) [deco]~/src/dandi-cli ❱ pytest -vvs dandi/tests/test_files.py::test_validate_bogus
=========================================================== test session starts ============================================================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
rootdir: /home/chymera/src/dandi-cli, configfile: tox.ini
plugins: pkgcore-0.12.18, mock-3.10.0
collected 1 item

dandi/tests/test_files.py::test_validate_bogus FAILED

================================================================= FAILURES =================================================================
___________________________________________________________ test_validate_bogus ____________________________________________________________
/usr/lib/python3.10/site-packages/_pytest/runner.py:339: in from_call
    result: Optional[TResult] = func()
/usr/lib/python3.10/site-packages/_pytest/runner.py:260: in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
/usr/lib/python3.10/site-packages/pluggy/_hooks.py:265: in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
/usr/lib/python3.10/site-packages/pluggy/_manager.py:80: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
/usr/lib/python3.10/site-packages/_pytest/unraisableexception.py:88: in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
/usr/lib/python3.10/site-packages/_pytest/unraisableexception.py:78: in unraisable_exception_runtest_hook
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E   pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7fbdd2b115a0>
E
E   Traceback (most recent call last):
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 758, in close_linked_files
E       for obj in self.__open_links:
E   AttributeError: 'NWBHDF5IO' object has no attribute '_HDF5IO__open_links'
E
E   During handling of the above exception, another exception occurred:
E
E   Traceback (most recent call last):
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/io.py", line 137, in __del__
E       self.close()
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 738, in close
E       self.close_linked_files()
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 766, in close_linked_files
E       warnings.warn("HDF5IO was not fully initialized before close. Missing self.__open_links.")
E   UserWarning: HDF5IO was not fully initialized before close. Missing self.__open_links.
=========================================================== slowest 10 durations ===========================================================
0.08s call     dandi/tests/test_files.py::test_validate_bogus
0.00s setup    dandi/tests/test_files.py::test_validate_bogus
0.00s teardown dandi/tests/test_files.py::test_validate_bogus
========================================================= short test summary info ==========================================================
FAILED dandi/tests/test_files.py::test_validate_bogus - pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7fbdd2b115a0>
============================================================ 1 failed in 0.23s =============================================================
```

@oruebel @rly 

Thanks so much for your help in sorting this out... however, it still fails:

```console
(dev) [deco]~/src/dandi-cli ❱ pytest -vvs dandi/tests/test_files.py::test_validate_bogus
=========================================================== test session starts ============================================================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0 -- /usr/bin/python3.10
cachedir: .pytest_cache
rootdir: /home/chymera/src/dandi-cli, configfile: tox.ini
plugins: pkgcore-0.12.18, mock-3.10.0
collected 1 item

dandi/tests/test_files.py::test_validate_bogus FAILED

================================================================= FAILURES =================================================================
___________________________________________________________ test_validate_bogus ____________________________________________________________
/usr/lib/python3.10/site-packages/_pytest/runner.py:339: in from_call
    result: Optional[TResult] = func()
/usr/lib/python3.10/site-packages/_pytest/runner.py:260: in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
/usr/lib/python3.10/site-packages/pluggy/_hooks.py:265: in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
/usr/lib/python3.10/site-packages/pluggy/_manager.py:80: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
/usr/lib/python3.10/site-packages/_pytest/unraisableexception.py:88: in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
/usr/lib/python3.10/site-packages/_pytest/unraisableexception.py:78: in unraisable_exception_runtest_hook
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
E   pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7f1f561d15a0>
E
E   Traceback (most recent call last):
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 760, in close_linked_files
E       for obj in self.__open_links:
E   AttributeError: 'NWBHDF5IO' object has no attribute '_HDF5IO__open_links'
E
E   During handling of the above exception, another exception occurred:
E
E   Traceback (most recent call last):
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/io.py", line 137, in __del__
E       self.close()
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 740, in close
E       self.close_linked_files()
E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 768, in close_linked_files
E       warnings.warn("HDF5IO was not fully initialized before close. Missing self.__open_links.")
E   UserWarning: HDF5IO was not fully initialized before close. Missing self.__open_links.
=========================================================== slowest 10 durations ===========================================================
0.08s call     dandi/tests/test_files.py::test_validate_bogus
0.00s setup    dandi/tests/test_files.py::test_validate_bogus
0.00s teardown dandi/tests/test_files.py::test_validate_bogus
========================================================= short test summary info ==========================================================
FAILED dandi/tests/test_files.py::test_validate_bogus - pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7f1f561d15a0>
============================================================ 1 failed in 0.23s =============================================================
(dev) [deco]~/src/dandi-cli ❱ pytest -vvs dandi/tests/test_files.py::test_validate_bogus &> /tmp/failpatch2
(dev) [deco]~/src/dandi-cli ❱ diff /tmp/failpatch /tmp/failpatch2
24c24
< E   pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7fec459c55a0>
---
> E   pytest.PytestUnraisableExceptionWarning: Exception ignored in: <function HDMFIO.__del__ at 0x7f976d7b95a0>
27c27
< E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 758, in close_linked_files
---
> E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 760, in close_linked_files
36c36
< E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 738, in close
---
> E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 740, in close
38c38
< E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 766, in close_linked_files
---
> E     File "/usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py", line 768, in close_linked_files
47c47
< ============================== 1 failed in 0.18s ===============================
---
> ============================== 1 failed in 0.19s ===============================
(dev) [deco]~/src/dandi-cli ❱ head -60 /usr/lib/python3.10/site-packages/hdmf/backends/hdf5/h5tools.py | tail -4

        self.__open_links = []  # keep track of other files opened from links in this file
        self.__file = None  # This will be set below, but set to None first in case an error occurs and we need to close
```

Last bit of the code showcasing that it's indeed the patch for the entire PR, this part being what https://github.com/hdmf-dev/hdmf/commit/f79b3a26bdfaee22b8d24bd3094c93be77e5595f adds.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,8 +1,8 @@
 # HDMF Changelog
 
-## HDMF 3.5.1 (Upcoming)
+## HDMF 3.5.1 (January 24, 2023)
 
-- Fixed bug when closing a ``HDF5IO`` on delete via the new ``HDMFIO.__del__`` before ``HDF5IO.__init__`` has been completed  @oruebel[#817](https://github.com/hdmf-dev/hdmf/pull/817)  
+- Fixed bug when closing a ``HDF5IO`` on delete via the new ``HDMFIO.__del__`` before ``HDF5IO.__init__`` has been completed. @oruebel [#817](https://github.com/hdmf-dev/hdmf/pull/817)  
 
 ## HDMF 3.5.0 (January 17, 2023)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 # HDMF Changelog
 
+## HDMF 3.5.1 (Upcoming)
+
+- Fixed bug when closing a ``HDF5IO`` on delete via the new ``HDMFIO.__del__`` before ``HDF5IO.__init__`` has been completed  @oruebel[#817](https://github.com/hdmf-dev/hdmf/pull/817)  
+
 ## HDMF 3.5.0 (January 17, 2023)
 
 ### Bug fixes

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -736,8 +736,15 @@ class HDF5IO(HDMFIO):
         """
         if close_links:
             self.close_linked_files()
-        if self.__file is not None:
-            self.__file.close()
+        try:
+            if self.__file is not None:
+                self.__file.close()
+        except AttributeError:
+            # Do not do anything in case that self._file does not exist. This
+            # may happen in case that an error occurs before HDF5IO has been fully
+            # setup in __init__, e.g,. if a child class such as NWBHDF5IO raises
+            # and error before self.__file has been created
+            warnings.warn("HDF5IO was not fully initialized before close. Missing self.__file")
 
     def close_linked_files(self):
         """Close all opened, linked-to files.
@@ -751,8 +758,12 @@ class HDF5IO(HDMFIO):
             for obj in self.__open_links:
                 if obj:
                     obj.file.close()
-        except AttributeError:  # In case that self.__open_links does not exist on delete
-            pass
+        except AttributeError:
+            # Do not do anything in case that self.__open_links does not exist. This
+            # may happen in case that an error occurs before HDF5IO has been fully
+            # setup in __init__, e.g,. if a child class such as NWBHDF5IO raises
+            # and error before self.__open_links has been created.
+            warnings.warn("HDF5IO was not fully initialized before close. Missing self.__open_links.")
         finally:
             self.__open_links = []
 

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -744,8 +744,8 @@ class HDF5IO(HDMFIO):
         except AttributeError:
             # Do not do anything in case that self._file does not exist. This
             # may happen in case that an error occurs before HDF5IO has been fully
-            # setup in __init__, e.g,. if a child class such as NWBHDF5IO raises
-            # and error before self.__file has been created
+            # setup in __init__, e.g,. if a child class (such as NWBHDF5IO) raises
+            # an error before self.__file has been created
             pass
 
     def close_linked_files(self):
@@ -763,8 +763,8 @@ class HDF5IO(HDMFIO):
         except AttributeError:
             # Do not do anything in case that self.__open_links does not exist. This
             # may happen in case that an error occurs before HDF5IO has been fully
-            # setup in __init__, e.g,. if a child class such as NWBHDF5IO raises
-            # and error before self.__open_links has been created.
+            # setup in __init__, e.g,. if a child class (such as NWBHDF5IO) raises
+            # an error before self.__open_links has been created.
             pass
         finally:
             self.__open_links = []

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -746,10 +746,15 @@ class HDF5IO(HDMFIO):
         not, which prevents the linked-to file from being deleted or truncated. Use this method to close all opened,
         linked-to files.
         """
-        for obj in self.__open_links:
-            if obj:
-                obj.file.close()
-        self.__open_links = []
+        # Make sure
+        try:
+            for obj in self.__open_links:
+                if obj:
+                    obj.file.close()
+        except AttributeError:  # In case that self.__open_links does not exist on delete
+            pass
+        finally:
+            self.__open_links = []
 
     @docval({'name': 'builder', 'type': GroupBuilder, 'doc': 'the GroupBuilder object representing the HDF5 file'},
             {'name': 'link_data', 'type': bool,

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -746,7 +746,7 @@ class HDF5IO(HDMFIO):
             # may happen in case that an error occurs before HDF5IO has been fully
             # setup in __init__, e.g,. if a child class (such as NWBHDF5IO) raises
             # an error before self.__file has been created
-            pass
+            self.__file = None
 
     def close_linked_files(self):
         """Close all opened, linked-to files.

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -55,6 +55,9 @@ class HDF5IO(HDMFIO):
         path, manager, mode, comm, file_obj, driver = popargs('path', 'manager', 'mode', 'comm', 'file', 'driver',
                                                               kwargs)
 
+        self.__open_links = []  # keep track of other files opened from links in this file
+        self.__file = None  # This will be set below, but set to None first in case an error occurs and we need to close
+
         if path is None and file_obj is None:
             raise ValueError("You must supply either a path or a file.")
 
@@ -89,7 +92,6 @@ class HDF5IO(HDMFIO):
         self.__dci_queue = HDF5IODataChunkIteratorQueue()  # a queue of DataChunkIterators that need to be exhausted
         ObjectMapper.no_convert(Dataset)
         self._written_builders = WriteStatusTracker()  # track which builders were written (or read) by this IO object
-        self.__open_links = []      # keep track of other files opened from links in this file
 
     @property
     def comm(self):

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -746,7 +746,7 @@ class HDF5IO(HDMFIO):
             # may happen in case that an error occurs before HDF5IO has been fully
             # setup in __init__, e.g,. if a child class such as NWBHDF5IO raises
             # and error before self.__file has been created
-            warnings.warn("HDF5IO was not fully initialized before close. Missing self.__file")
+            pass
 
     def close_linked_files(self):
         """Close all opened, linked-to files.
@@ -765,7 +765,7 @@ class HDF5IO(HDMFIO):
             # may happen in case that an error occurs before HDF5IO has been fully
             # setup in __init__, e.g,. if a child class such as NWBHDF5IO raises
             # and error before self.__open_links has been created.
-            warnings.warn("HDF5IO was not fully initialized before close. Missing self.__open_links.")
+            pass
         finally:
             self.__open_links = []
 

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -833,13 +833,11 @@ class TestHDF5IO(TestCase):
         """
         class MyHDF5IO(HDF5IO):
             def __init__(self):
-                raise ValueError("test error")
-        with self.assertWarnsWith(warn_type=UserWarning,
-                                  exc_msg="HDF5IO was not fully initialized before close. Missing self.__file"):
-            try:
-                with MyHDF5IO() as _:
-                    pass
-            except ValueError:
+                self.__open_links = []
+                raise ValueError("interrupt before HDF5IO.__file is initialized")
+
+        with self.assertRaisesWith(exc_type=ValueError, exc_msg="interrupt before HDF5IO.__file is initialized"):
+            with MyHDF5IO() as _:
                 pass
 
     def test_delete_with_incomplete_construction_missing_open_files(self):
@@ -847,17 +845,13 @@ class TestHDF5IO(TestCase):
         Here we test what happens when close is called before HDF5IO.__init__ has
         bee completed. In this case self.__open_file is missing
         """
-
         class MyHDF5IO(HDF5IO):
             def __init__(self):
-                raise ValueError("test error")
+                self.__file = None
+                raise ValueError("interrupt before HDF5IO.__open_files is initialized")
 
-        with self.assertWarnsWith(warn_type=UserWarning,
-                                  exc_msg="HDF5IO was not fully initialized before close. Missing self.__open_links."):
-            try:
-                with MyHDF5IO() as _:
-                    pass
-            except ValueError:
+        with self.assertRaisesWith(exc_type=ValueError, exc_msg="interrupt before HDF5IO.__open_files is initialized"):
+            with MyHDF5IO() as _:
                 pass
 
     def test_set_file_mismatch(self):

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -826,6 +826,40 @@ class TestHDF5IO(TestCase):
             self.assertEqual(io.manager, self.manager)
             self.assertEqual(io.source, self.path)
 
+    def test_delete_with_incomplete_construction_missing_file(self):
+        """
+        Here we test what happens when close is called before HDF5IO.__init__ has
+        bee completed. In this case self.__file is missing
+        """
+        class MyHDF5IO(HDF5IO):
+            def __init__(self):
+                raise ValueError("test error")
+        with self.assertWarnsWith(warn_type=UserWarning,
+                                  exc_msg="HDF5IO was not fully initialized before close. Missing self.__file"):
+            try:
+                with MyHDF5IO() as _:
+                    pass
+            except ValueError:
+                pass
+
+    def test_delete_with_incomplete_construction_missing_open_files(self):
+        """
+        Here we test what happens when close is called before HDF5IO.__init__ has
+        bee completed. In this case self.__open_file is missing
+        """
+
+        class MyHDF5IO(HDF5IO):
+            def __init__(self):
+                raise ValueError("test error")
+
+        with self.assertWarnsWith(warn_type=UserWarning,
+                                  exc_msg="HDF5IO was not fully initialized before close. Missing self.__open_links."):
+            try:
+                with MyHDF5IO() as _:
+                    pass
+            except ValueError:
+                pass
+
     def test_set_file_mismatch(self):
         self.file_obj = File(get_temp_filepath(), 'w')
         err_msg = ("You argued %s as this object's path, but supplied a file with filename: %s"

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -828,8 +828,8 @@ class TestHDF5IO(TestCase):
 
     def test_delete_with_incomplete_construction_missing_file(self):
         """
-        Here we test what happens when close is called before HDF5IO.__init__ has
-        bee completed. In this case self.__file is missing
+        Here we test what happens when `close` is called before `HDF5IO.__init__` has
+        been completed. In this case, self.__file is missing.
         """
         class MyHDF5IO(HDF5IO):
             def __init__(self):
@@ -842,8 +842,8 @@ class TestHDF5IO(TestCase):
 
     def test_delete_with_incomplete_construction_missing_open_files(self):
         """
-        Here we test what happens when close is called before HDF5IO.__init__ has
-        bee completed. In this case self.__open_file is missing
+        Here we test what happens when `close` is called before `HDF5IO.__init__` has
+        been completed. In this case, self.__open_files is missing.
         """
         class MyHDF5IO(HDF5IO):
             def __init__(self):

